include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -525,337 +544,790 @@ namespace internal
+     */
+    static void do_mass (const AlignedVector<Number2> &transformation_matrix,
+                         const AlignedVector<Number>  &coefficients,
+                         const Number *values_in,
+                         Number *scratch_data,
+                         Number *values_out)
+    {
+      constexpr int next_dim = dim > 1 ? dim-1 : dim;
+      Number *my_scratch = n_points_1 != n_points_2 ? scratch_data : values_out;
+      for (unsigned int q=n_points_1; q!=0; --q)
+        FEEvaluationImplBasisChange<variant,next_dim,n_points_1,n_points_2,n_components,Number,Number2>
+        ::do_forward(transformation_matrix,
+                     values_in + (q-1)*Utilities::fixed_int_power<n_points_1,dim-1>::value,
+                     my_scratch + (q-1)*Utilities::fixed_int_power<n_points_2,dim-1>::value);
+      EvaluatorTensorProduct<variant, dim, n_points_1, n_points_2,
+                             Number,Number2> eval_val (transformation_matrix);
+      const unsigned int n_inner_blocks = (dim > 1 && n_points_2 < 10) ? n_points_2 : 1;
+      const unsigned int n_blocks = Utilities::fixed_int_power<n_points_2,dim-1>::value;
+      for (unsigned int ii=0; ii<n_blocks; ii+=n_inner_blocks)
+        for (unsigned int c=0; c<n_components; ++c)
+          {
+            for (unsigned int i=ii; i<ii+n_inner_blocks; ++i)
+              eval_val.template values_one_line<dim-1,true,false> (my_scratch+i, my_scratch+i);
+            for (unsigned int q=0; q<n_points_2; ++q)
+              for (unsigned int i=ii; i<ii+n_inner_blocks; ++i)
+                my_scratch[i+q*n_blocks] *= coefficients[i+q*n_blocks];
+            for (unsigned int i=ii; i<ii+n_inner_blocks; ++i)
+              eval_val.template values_one_line<dim-1,false,false>(my_scratch+i, my_scratch+i);
+          }
+      for (unsigned int q=0; q<n_points_1; ++q)
+        FEEvaluationImplBasisChange<variant,next_dim,n_points_1,n_points_2,n_components,Number,Number2>
+        ::do_backward(transformation_matrix, false,
+                      my_scratch + q*Utilities::fixed_int_power<n_points_2,dim-1>::value,
+                      values_out + q*Utilities::fixed_int_power<n_points_1,dim-1>::value);
+    }
+  };
+
+
+
+  /**
+   * This struct performs the evaluation of function values, gradients and
+   * Hessians for tensor-product finite elements. This a specialization for
+   * elements where the nodal points coincide with the quadrature points like
+   * FE_Q shape functions on Gauss-Lobatto elements integrated with
+   * Gauss-Lobatto quadrature. The assumption of this class is that the shape
+   * 'values' operation is identity, which allows us to write shorter code.
+   *
+   * In literature, this form of evaluation is often called spectral
+   * evaluation, spectral collocation or simply collocation, meaning the same
+   * location for shape functions and evaluation space (quadrature points).
+   *
+   * @author Katharina Kormann, 2012
```

Comment:
you have been sitting on this since 2012? ;-)

Reply:
No, it is just github that gets confused by moving this class `FEEvaluationImplCollocation` above `FEEvaluationImplTransformToCollocation` - the latter now uses the former after having changed the basis, as promised by the name.

And just as a side note, the changes in that file are actually mostly 2017.